app/code/Magento/CatalogUrlRewrite/Model/Product/Validator.php

```diff
@@ -0,0 +1,97 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\CatalogUrlRewrite\Model\Product;
+
+use Magento\Catalog\Model\Product;
+use Magento\CatalogUrlRewrite\Model\ProductUrlPathGenerator;
+use Magento\UrlRewrite\Model\UrlFinderInterface;
+use Magento\UrlRewrite\Service\V1\Data\UrlRewrite;
+use Magento\CatalogUrlRewrite\Model\ProductUrlRewriteGenerator;
+use Magento\Store\Model\StoreManagerInterface;
+
+/**
+ * Url Rewrites Product validator.
+ */
+class Validator
+{
+    /**
+     * @var ProductUrlPathGenerator
+     */
+    private $productUrlPathGenerator;
+
+    /**
+     * @var UrlFinderInterface
+     */
+    private $urlFinder;
+
+    /**
+     * @var StoreManagerInterface
+     */
+    private $storeManager;
+
+    /**
+     * @param ProductUrlPathGenerator $productUrlPathGenerator
+     * @param UrlFinderInterface $urlFinder
+     * @param StoreManagerInterface $storeManager
+     */
+    public function __construct(
+        ProductUrlPathGenerator $productUrlPathGenerator,
+        UrlFinderInterface $urlFinder,
+        StoreManagerInterface $storeManager
+    ) {
+        $this->productUrlPathGenerator = $productUrlPathGenerator;
+        $this->urlFinder = $urlFinder;
+        $this->storeManager = $storeManager;
+    }
+
+    /**
+     * Find Url Key conflicts of a product.
+     *
+     * @param Product $product
+     * @return array Array of conflicting Url Keys.
+     */
+    public function findUrlKeyConflicts(Product $product): array
+    {
+        if (!$product->getUrlKey()) {
+            $urlKey = $this->productUrlPathGenerator->getUrlKey($product);
+            $product->setUrlKey($urlKey);
```

Comment:
I suppose this UrlKey generation if t empty should not be in validator.

Reply:
Thank you, fixed.